modules/optimization/src/executioners/AdjointTransientSolve.C

```diff
@@ -0,0 +1,118 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdjointTransientSolve.h"
+
+#include "FEProblemBase.h"
+#include "NonlinearSystemBase.h"
+#include "MooseUtils.h"
+
+#include "libmesh/sparse_matrix.h"
+#include "libmesh/numeric_vector.h"
+
+InputParameters
+AdjointTransientSolve::validParams()
+{
+  InputParameters params = AdjointSolve::validParams();
+  return params;
+}
+
+AdjointTransientSolve::AdjointTransientSolve(Executioner & ex)
+  : AdjointSolve(ex), _old_time_residual(_nl_adjoint.getResidualTimeVector())
+{
+}
+
+bool
+AdjointTransientSolve::solve()
+{
+  bool converged = AdjointSolve::solve();
+
+  // Gather the contribution of this timestep to add to the next solve's source
+  if (converged)
+    evaluateTimeResidual(_nl_adjoint.solution(), _old_time_residual);
+
+  return converged;
+}
+
+void
+AdjointTransientSolve::insertForwardSolution(int tstep)
+{
+  // Time step should not be negative
+  if (tstep < 0)
+    mooseError("Negative time step occurred.");
+  std::size_t t_step = tstep; // Avoid compiler warnings
+  // Should not be inserting a time greater the last one inserted
+  if (t_step > _forward_solutions.size())
+    mooseError("Trying to insert a solution that is greater than one post the previously inserted "
+               "time step.");
+  // Add vector to the adjoint system if this time has not occurred
+  else if (t_step == _forward_solutions.size())
+  {
+    std::string name = "forward_solution_" + std::to_string(t_step);
+    _forward_solutions.push_back(&_nl_adjoint.addVector(name, false, PARALLEL));
+  }
+
+  // Set the vector to the inserted solution
+  *_forward_solutions[t_step] = _nl_forward.solution();
+}
+
+void
+AdjointTransientSolve::setForwardSolution(int tstep)
+{
+  // Make sure the time step was saved
+  if (tstep < 0 || tstep >= (int)_forward_solutions.size())
+    mooseError("Could not find forward solution at time step ", tstep, ".");
+  std::size_t t_step = tstep; // Avoid compiler warnings
```

Comment:
The `_t_step` member variable in Transient (which is what is passed into here) is `int`. It should never be negative, but I want to throw an error just in case.